control-plane-operator/controllers/hostedcontrolplane/dnsoperator/dnsoperator.go

```diff
@@ -0,0 +1,195 @@
+		},
+		ReleaseVersion:          version,
+		AvailabilityProberImage: images[util.AvailabilityProberImageName],
+		OwnerRef:                config.OwnerRefFrom(hcp),
+	}
+
+	p.DeploymentConfig.AdditionalAnnotations = map[string]string{
+		"target.workload.openshift.io/management": `{"effect": "PreferredDuringScheduling"}`,
+	}
+	p.DeploymentConfig.AdditionalLabels = map[string]string{
+		"name":                        "dns-operator",
+		hyperv1.ControlPlaneComponent: "dns-operator",
+	}
+	p.DeploymentConfig.Scheduling.PriorityClass = config.DefaultPriorityClass
+	p.DeploymentConfig.SetRestartAnnotation(hcp.ObjectMeta)
+	p.DeploymentConfig.SetDefaults(hcp, nil, utilpointer.IntPtr(1))
+	p.DeploymentConfig.SetDefaultSecurityContext = setDefaultSecurityContext
+
+	return p
+}
+
+// ReconcileDeployment reconciles a deployment of the DNS operator, which runs
+// in the management cluster and manages operands in a hosted cluster.  For
+// non-HyperShift clusters, the DNS operator is deployed by
+// cluster-version-operator with the following manifest:
+// <https://github.com/openshift/cluster-dns-operator/blob/master/manifests/0000_70_dns-operator_02-deployment.yaml>.
+// For HyperShift, the deployment differs from non-HyperShift clusters in the
+// following ways:
+//
+// * The operator is configured with a kubeconfig for the managed cluster.
+//
+// * The operator metrics are exposed using cleartext rather than being
+//   protected using kube-rbac-proxy.  (However, CoreDNS's metrics are still
+//   protected using kube-rbac-proxy on the hosted cluster.)
+//
+// * The operator has HyperShift-specific annotations, labels, owner reference,
+//   and affinity rules and omits the node selector for control-plane nodes.
+//
+// * The operator has an init container to delete any deployment that is running
+//   in the hosted cluster.  This is to avoid having multiple, conflicting
+//   operator deployments when HyperShift is updated from a version that allowed
+//   the DNS operator to run in the hosted cluster to a version of HyperShift
+//   that runs the DNS operator in the management cluster.
+//
+// * The operator has an additional init container that probes the hosted
+//   cluster's kube-apiserver to verify that the dnses.operator.openshift.io API
+//   is available.
+//
+// The DNS operator does not require access to the cloud platform API,
+// hosted-cluster services, or external services, so the operator does not
+// require any special proxy configuration or permissions in the management
+// cluster.
+func ReconcileDeployment(dep *appsv1.Deployment, params Params, apiPort *int32) {
+	params.OwnerRef.ApplyTo(dep)
+	dep.Spec.Selector = &metav1.LabelSelector{
+		MatchLabels: map[string]string{"name": "dns-operator"},
+	}
+	dep.Spec.Template.Spec.AutomountServiceAccountToken = utilpointer.BoolPtr(false)
+	dep.Spec.Template.Spec.InitContainers = []corev1.Container{{
+		// Delete any existing dns-operator deployment in the managed
```

Comment:
I'm curious why you choose init container vs doing this programatically?

Reply:
To do it programmatically, would the appropriate thing be to add a `DeleteIfNeeded` call near the end of `(r *HostedClusterReconciler) reconcile`?

Reply:
> To do it programmatically, would the appropriate thing be to add a DeleteIfNeeded call near the end of (r *HostedClusterReconciler) reconcile?

No, the hostedclusterreconciler doesn't reconcile things inside the hostedcluster. That would be in the `hostedclusterconfigoperator`, so adding a `deleteIfNeeded` for these resources somewhere in control-plane-operator/hostedclusterconfigoperator/controllers/resources/